### ISSUE TITLE
`handler and runtime must be set when PackageType is Zip`

### ISSUE THREAD

**Issue by arpadgabor** [sst/ion#800]
```
diffing urn:pulumi:arpad::nolas::sst:aws:Function$aws:lambda/function:Function::CrmApiFunction:
  handler and runtime must be set when PackageType is Zip
```

We are currently on 0.1.29 due to this error (and the ports one). We have about 8 people on the team working with their local stages. We could get them to do `sst remove` and then `sst dev` again but I saw this error happening before on other versions on Discord - so we'd like to upgrade whenever we are sure in some way that future upgrades don't require removing infra (wouldn't be that big of a problem if cloudfront wouldn't take so long to create or delete).

Let me know if I can provide anything extra, I checked the state file but I think it's the old one not the state it's trying to deploy, so there wasn't anything interesting there.

**Comment by arpadgabor** [comment]
Oh and I tried running `diff` but I get the same error before generating any diffs.

**Comment by jayair** [comment]
Just to clarify, which version are you getting this error with?

**Comment by arpadgabor** [comment]
0.1.76 or whatever was latest when I posted, but happens since some version from last week (can't tell exactly right now as I'm on mobile)

**Comment by srodriki** [comment]
I'm having the exact same issue as @arpadgabor 
It's happening to me with 0.1.76 and was running fine a few days ago with a previous version. Just FYI, runtime and handler are indeed set from my infrastructure code.

**Comment by fwang** [comment]
Thanks @srodriki. I might need some help tracking this down.

When u run into again, can u send me a few things to frank@sst.dev
- a screenshot of the terminal output
- zip up `.sst/.pulumi/stacks` folder
- zip up `.sst/log` folder

**Comment by srodriki** [comment]
Hey @fwang sure thing, I'll send everything if it happens again. I discussed this issue with @thdxr in private and shared the pulumi folder with him.
In the end and by Dax's advice, I deleted the whole stage and started from scratch to fix the issue because it was honestly hard to pinpoint.

**Comment by 0cv** [comment]
@fwang I've just had this issue and just sent to you the requested files. 

I didn't see how to resolve this issue either, despite the name being pretty clear, and which lambda function was impacted - I commented out the part of this stack referencing the name and still the error was there. The only solution I found was indeed to delete the stack and re-creating it (which has some impacts because the Cognito Pools etc. are all different and this breaks the SSO etc.).

This happened when upgrading from 0.1.40 to the latest version as of today (0.1.87)

**Comment by arpadgabor** [comment]
I found a possible partial solution but it still involves removing resources:

```
sst remove --target "**sst:aws:Function::**" --stage {stage}
```

Be aware that if you have Routers that have a function as dependency, it will delete the router too. But at least you don't delete databases.

**Comment by fwang** [comment]
@0cv i didn't see anything weird weird in the files. I also tried creating a new app in 0.1.40 with similar components i saw in ur state files, then redeploy using 0.1.87.. no luck.

Just to clarify, u can redeploy using 0.1.40. And deploy only fails if u update to 0.1.87?

I might need some help pinpoint the exact version they wasn't able to upgrade to.

**Comment by 0cv** [comment]
@fwang that's right. I don't have the error anymore because I deleted the entire stack and recreated it. I must also say that I didn't create the stack with 0.1.40, it was created with a previous version (I don't know which one anymore) - maybe this is important? My stack is also relatively complicated, I have from 10 to 15 lambdas, a cloudfront distribution, an ECS service, 2 API gateways, login with Cognito ... The error was clearly pointing to a specific Lambda but deleting it from the stack was still showing this error, it's as much as I can say ... Sorry I can't help more on that but if that happens again, I will try to provide more info

**Comment by fwang** [comment]
Managed to track down the issue. Should be fixed in the latest release.

Feel free to re-open if issue still happens.

**Comment by ryands17** [comment]
@fwang This still happens on `sst diff`. `sst deploy` works fine.

SST version: 3.0.70

**Comment by arpadgabor** [comment]
On vacation now but can confirm it happened to me as well last week 😂

**Comment by jayair** [comment]
Uh oh

**Comment by jayair** [comment]
Can you run this with `--print-logs`?

**Comment by veljkoza** [comment]
So I managed to reset it with:
```
sst remove --target "**sst:aws:Function::**" --stage {stage}
```

Thanks @arpadgabor , although the reinstallation of resources took around 10 mintues

**Comment by fwang** [comment]
Pushed out a fix in v3.1.18.

I'm pretty confident this fixes it. If not, let's reopen this.

Thanks everyone!

**Comment by bigbyte-dom** [comment]
@fwang unfortunately, still running into the same issue on 3.1.18. 

I'll fire over some logs.


```
|  Error       UIDefaultFunction aws:lambda:Function
|    sdk-v2/provider2.go:472: sdk.helper_schema: handler and runtime must be set when PackageType is Zip: provider=aws@6.51.1
|  Error       AppSyncApi aws:appsync:GraphQLApi
|  handler and runtime must be set when PackageType is Zip
```

- Deleting the AppSync lambda functions -> refresh -> deploy = same error :(
- Deleting the AppSync Domain and API -> refresh -> deploy = success!!

We've had the issue pop up previously on NextJS DefaultFunction too, but not this time.

It's appeared on every build/stage as it bubbled up from local -> PR -> dev -> prod. Each stage we had to delete the same resources.

We had also set runtime 'nodejs20.x. on both the UI transforms and AppSync functions with the same error.

**Comment by san4d** [comment]
I'm also still having this issue, even after upgrading to `3.1.22`:
```text
|    sdk-v2/provider2.go:457: sdk.helper_schema: handler and runtime must be set when PackageType is Zip: provider=aws@6.51.0
|  Error
|  handler and runtime must be set when PackageType is Zip

✕  Failed
   CoreDevServerFunction aws:lambda:Function
     sdk-v2/provider2.go:457: sdk.helper_schema: handler and runtime must be set when PackageType is Zip: provider=aws@6.51.0
   handler and runtime must be set when PackageType is Zip
```

**Comment by oyed** [comment]
also seeing this quite often now

**Comment by fwang** [comment]
hmm.. does this happen right after a failed `sst deploy`, failed `sst remove`, or `sst sync`?

If anyone is running into this w/ a reproducible setup at hand and want to live debug, text me 6478061629!!!

**Comment by fwang** [comment]
I need some help tracking this down.

When someone run get this error, can u send the following to my email [frank@sst.dev](mailto:frank@sst.dev)

- full console output (including the version line up top)
- zip up .sst/.pulumi/stacks folder
- zip up .sst/log folder

**Comment by san4d** [comment]
Thanks for the reply. I'll try to reproduce tomorrow. It happened to me during standard dev reloading.

**Comment by kaceycleveland** [comment]
Not sure if it helps but I ran into this when I had a SQS queue with a hooked up lambda subscribed to it. It was deployed with the dev server running. I then changed the SQS queue to `fifo` and it started complaining and failed to redeploy.

**Comment by arpadgabor** [comment]
Somehow managed to reproduce the error when a dependency of an `sst.Linkable` changed. We are generating various SSM parameters and we keep the ARN inside linkables. I wanted to restructure the parameters and got the error (in dev mode). Will try to make a reproduction some day, but until I get the time I wanted to let you know this. Reverted the changes and it's fixed back.

**Comment by fbahacker21** [comment]
Got the same error today, the only thing happened to me is I change my aws SQS queue to fifo: true

**Comment by danew** [comment]
Ran into this today on `3.1.63`, both in dev and production. It happens if you edit a function source while deploying

**Comment by jayair** [comment]
That's an interesting hint @fwang

**Comment by fwang** [comment]
Quick update.. we managed to track down the issue.

Working on a fix. Will keep everyone posted.
